File: core/src/com/t2wonderland/kurona/Objects/Kurona.java
package com.t2wonderland.kurona.Objects;

import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.t2wonderland.kurona.Interfaces.CharacterState;
import com.t2wonderland.kurona.Interfaces.ICharacterObject;

public class Kurona implements ICharacterObject {

	private float VELOCITY_DASH = 10f;
	private float VELOCITY_SLOW = 2.5f;
	private float VELOCITY_SP = 20f;

	private float WIDTH = 2f;
	private float HEIGHT = 2f;

	private Vector2 _size = new Vector2(WIDTH, HEIGHT);
	private Vector2 _position;
	private Rectangle _bounds;
	private Vector2 _velocity;
	private CharacterState _state;
	private float _stateTime;
	private float _maxVelocityX, _maxVelocityY;
	
	public Kurona() {
		_state = CharacterState.Slow;
		_stateTime = 0;

		// starting at the slow mode
		_velocity = new Vector2();
		_velocity.x = VELOCITY_SLOW;
		_maxVelocityX = VELOCITY_DASH;
		_maxVelocityY = VELOCITY_DASH;
	}

	@Override
	public void updatePosition(float deltaTime, Vector2 accelValue) {
		updateVelocity(deltaTime, accelValue);
		updateState(deltaTime);
		updatePositionAndBounds(deltaTime);
	}

	private void updateVelocity(float deltaTime, Vector2 accelValue) {
		Vector2 velocity = calculateVelocity(deltaTime, accelValue);
		_velocity.add(velocity);
		adjustVelocity();
	}

	private Vector2 calculateVelocity(float deltaTime, Vector2 accelValue) {
		Vector2 velocity = new Vector2();
		if (_maxVelocityX < _velocity.x){
			velocity.x = 0;
		}
		else{
			velocity.x = accelValue.x * deltaTime;
		}

		if (_maxVelocityY < _velocity.y){
			velocity.y = 0;
		}
		else{
			velocity.y = accelValue.y * deltaTime;
		}

		return velocity;
	}

	private void adjustVelocity() {
		if(_maxVelocityX < _velocity.x)
		{
			_velocity.x = _maxVelocityX;
		}

		if (_maxVelocityY<_velocity.y)
		{
			_velocity.y = _maxVelocityY;
		}
	}

	private void updateState(float deltaTime) {
		if (_state == CharacterState.Slow){
			_stateTime += deltaTime/2;
		}
		else{
			_stateTime += deltaTime;
		}
	}

	private void updatePositionAndBounds(float deltaTime) {
		_position.add(_velocity.x * deltaTime, _velocity.y * deltaTime);
		_bounds.x = _position.x - _bounds.width / 2;
		_bounds.y = _position.y - _bounds.height / 2;
	}

	@Override
	public CharacterState getCharacterState() {
		return _state;
	}

	@Override
	public float getStateTime() {
		return _stateTime;
	}

	@Override
	public void hitBarricade() {
		_velocity.x = VELOCITY_SLOW;
		_state = CharacterState.Slow;
	}

	@Override
	public void releaseBarricade() {
		_velocity.x = VELOCITY_DASH;
		_state = CharacterState.Dash;
	}

	@Override
	public void setInitialPosition(float x, float y) {
		this._position = new Vector2(x, y);
		this._bounds = new Rectangle(x - WIDTH / 2, y - HEIGHT / 2, WIDTH, HEIGHT);
	}

	@Override
	public Vector2 getSize() {
		return _size;
	}

	@Override
	public void setPosition(Vector2 position) {
		_position = position;
	}

	@Override
	public Vector2 getPosition() {
		return _position;
	}

	@Override
	public Rectangle getBounds() {
		return _bounds;
	}
}

File: core/src/com/t2wonderland/kurona/Objects/Kurona.kt
package com.t2wonderland.kurona.Objects

import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.t2wonderland.kurona.Interfaces.CharacterState
import com.t2wonderland.kurona.Interfaces.ICharacterObject

class Kurona : ICharacterObject {

    private val VELOCITY_DASH = 10f
    private val VELOCITY_SLOW = 2.5f
    private val VELOCITY_SP = 20f

    private val WIDTH = 2f
    private val HEIGHT = 2f

    override val size = Vector2(WIDTH, HEIGHT)
    override var position: Vector2 = Vector2()
    override var bounds: Rectangle = Rectangle()
        private set
    private val _velocity: Vector2
    override var characterState: CharacterState = CharacterState.Dash
        private set
    override var stateTime: Float = 0.toFloat()
        private set
    private val _maxVelocityX: Float
    private val _maxVelocityY: Float

    init {
        characterState = CharacterState.Slow
        stateTime = 0f

        // starting at the slow mode
        _velocity = Vector2()
        _velocity.x = VELOCITY_SLOW
        _maxVelocityX = VELOCITY_DASH
        _maxVelocityY = VELOCITY_DASH
    }

    override fun updatePosition(deltaTime: Float, accelValue: Vector2) {
        updateVelocity(deltaTime, accelValue)
        updateState(deltaTime)
        updatePositionAndBounds(deltaTime)
    }

    private fun updateVelocity(deltaTime: Float, accelValue: Vector2) {
        val velocity = calculateVelocity(deltaTime, accelValue)
        _velocity.add(velocity)
        adjustVelocity()
    }

    private fun calculateVelocity(deltaTime: Float, accelValue: Vector2): Vector2 {
        val velocity = Vector2()
        if (_maxVelocityX < _velocity.x) {
            velocity.x = 0f
        } else {
            velocity.x = accelValue.x * deltaTime
        }

        if (_maxVelocityY < _velocity.y) {
            velocity.y = 0f
        } else {
            velocity.y = accelValue.y * deltaTime
        }

        return velocity
    }

    private fun adjustVelocity() {
        if (_maxVelocityX < _velocity.x) {
            _velocity.x = _maxVelocityX
        }

        if (_maxVelocityY < _velocity.y) {
            _velocity.y = _maxVelocityY
        }
    }

    private fun updateState(deltaTime: Float) {
        if (characterState === CharacterState.Slow) {
            stateTime += deltaTime / 2
        } else {
            stateTime += deltaTime
        }
    }

    private fun updatePositionAndBounds(deltaTime: Float) {
        position!!.add(_velocity.x * deltaTime, _velocity.y * deltaTime)
        bounds!!.x = position!!.x - bounds!!.width / 2
        bounds!!.y = position!!.y - bounds!!.height / 2
    }

    override fun hitBarricade() {
        _velocity.x = VELOCITY_SLOW
        characterState = CharacterState.Slow
    }

    override fun releaseBarricade() {
        _velocity.x = VELOCITY_DASH
        characterState = CharacterState.Dash
    }

    override fun setInitialPosition(x: Float, y: Float) {
        this.position.x = x
        this.position.y = y

        this.bounds.x = x - WIDTH / 2
        this.bounds.y = y - HEIGHT / 2
        this.bounds.width = WIDTH
        this.bounds.height = HEIGHT
    }
}
